src/neo/IO/Caching/DataCache.cs

```diff
@@ -47,8 +47,17 @@ public void Add(TKey key, TValue value)
         {
             lock (dictionary)
             {
-                if (dictionary.TryGetValue(key, out Trackable trackable) && trackable.State != TrackState.Deleted)
-                    throw new ArgumentException();
+                if (dictionary.TryGetValue(key, out Trackable trackable))
+                {
+                    if (trackable.State != TrackState.Deleted)
+                        throw new ArgumentException();
+                }
+                else
+                {
+                    if (TryGetInternal(key) != null)
+                        throw new ArgumentException();
+                }
```

Comment:
When we are using `Add()`, we assume that the data does not exist. If we can't assume that, we should use `GetAndChange()`. So, since we are using `Add()`, there is no need to add another data read operation.

Reply:
The idea behind this PR (and the above code change) is create consistent behaviour of the cache by not just validating that the value exists in cache, but to also validate that it does not exist in the real backend. You previously choose this exact solution (https://github.com/neo-project/neo/issues/1173#issuecomment-544890743) instead of always overriding the data (described in the linked comment as option 1).

If we don't choose either option 1 or 2 we will have inconsistent behaviour as described in #1173

I don't mind which of the 2 we choose, but right now the feedback looks like you don't want to change anything?! 😕 I think it is wrong to **assume** that data does not exist and I think you agreed with that when you wrote the code initially, otherwise this check would not exists
```C#
    if (trackable.State != TrackState.Deleted)
                        throw new ArgumentException(); 
```

Reply:
The overhead of this check is very small, so just do it by the way.

Reply:
I did not understand the point as well, @erikzhang.

The implementation here seems to be option 2 of that aforementioned issue.
Do you want to leave the code as in the original implementation?

Reply:
I don't want to change it. The purpose of `Add()` is to add records when they do not exist, and save the cost of a read operation. If you are not sure that the record exists or not, you should use `GetAndChange()`.

Reply:
I see, @erikzhang.
It looks like to be more efficient in the original code, yes? In the sense of avoiding a second check with `if (TryGetInternal(key) != null)`.

@ixje, if that is the case, perhaps it is better to close PR and issue and open another one for commenting the code properly in order to avoid future misinterpretations.

Reply:
So everybody basically wasted their time in creating the PR, reviewing it and fixing unit tests (still thanks for that @vncoelho) because 1.5 months after opening the PR and the initial review we're suddenly trying to save performance on a db operation that takes < 10 microseconds (according to leveldb benchmarks for random reads) while elsewhere in the code base millisecond delays are accepted.

Don't get me wrong here. I'm all cool with keeping performance in mind for reaching the highest TPS, but how are we supposed to make a sound decision what makes sense to spend time on when performance criteria seems randomly applied?

Reply:
@ixje, I really apologize for that.

I really did not have enough knowledge for checking it before.
In fact, I would like to thank you for your dedication. In addition, your insights helped me to investigate further.

Perhaps, there are methods that  are not being used from the `DataCache`, such as `GetOrAdd`.
We need to solve these questions and improve what needs to be improved.